Improved perfomance

it1804 opened the pull request:
The marshalOID () and oidToString () functions use the string functions Trim, Split, Join, which work internally with an array of strings. Also, to convert the oid string to an array of numbers and in the opposite direction, an additional array is created inside both marshalOID () and oidToString (), which causes at least twice the load on the GC.
In my case, with parallel polling of 100k + devices, this change in these functions improved performance and CPU load by 10% -15%. The average CPU load dropped from 80% to 65% after. In addition, the total polling time has noticeably decreased.

The second patch (reverted due to broken ci check) concerns the logging function logger.Printf(). Even with logging disabled, the Printf() system call is still called, which, according to pprof, also consumes a significant portion of the CPU time.
I don't know what to do in this case, so I just commented them out in problem areas. Logically, the logPrintf () function should be called here, but in these cases it is not available for call.

TimRots commented:
Thank you for another great PR @it1804.

Perhaps it's better to split-off the second patch you mention into a separate PR?
I btw do see some overlap with #313, please feel free to join in the conversation there if you think the issue you mention can be fixed in parallel. Otherwise opening a separate issue to untangle things would be nice to have.

it1804 commented:
> Perhaps it's better to split-off the second patch you mention into a separate PR? As I do see some overlap with #313 please feel free to join in the conversation there if you think the issue you mention can be fixed in parallel.
> Otherwise opening a separate issue to untangle things would be nice to have.

Yes, I will join that discussion. I think that both logging problems can be solved in one place. The second patch shouldn't have been included in this pr.

TimRots commented on helper.go:
```diff
@@ -463,24 +463,45 @@ func marshalObjectIdentifier(oid []int) (ret []byte, err error) {
+		if m < 0 {
+			break
+		}
+		a[i], err = strconv.Atoi(oid[:m])
```
if an error occurs during ``strconv.Atoi(oid[:m])``, an error in the form of ``unable to marshal OID: unable to parse OID: value out of range`` is displayed by the calling ``marshalOID()`` function, which is a bit too verbose perhaps. Can we handle this in a nicer way?

it1804 replied:
I also agree. Changed the error text

TimRots commented on helper.go:
```diff
@@ -463,24 +463,45 @@ func marshalObjectIdentifier(oid []int) (ret []byte, err error) {
 
+func marshalOID(oid string) ([]byte, error) {
+	var err error
+
```
If we are touching this function should we also consider zooming out a bit, and consider the following:
* adding an empty oid string check, and return an error if detected?
* check if OID length >= length as defined in [rfc2578](https://tools.ietf.org/html/rfc2578#section-7.1.3), and return an error if detected?

it1804 replied:
Added validation of sub-identifiers according to clause 3.5 [rfc2578](https://tools.ietf.org/html/rfc2578#section-3.5) , as well as checking for an empty string

TimRots replied:
Like this approach a lot. Thank you

TimRots reviewed (changes requested):
Added some initial comments on your PR @it1804 
Would it be possible to share code that does an actual benchmark of your change,
so we can perhaps improve it further?

it1804 commented:
My tests are simple. The functions are called in a loop with one million iterations.
In the case of oidToString, I additionally tested one more implementation using a byte buffer, it is close to my result, but slightly worse.
For the oidToIntArray function, I took the marshalOID part, in which the string is directly converted into an array of numbers.
You need to run it locally on your computer, The Go Playground does not allow you to run the program with a long timeout. 

[Benchmark oidToIntArray](https://play.golang.org/p/B_TTWDZlDl_m)
[Benchmark oidToString](https://play.golang.org/p/MQvzWmRVHi6)

it1804 commented on helper.go:
```diff
@@ -463,24 +463,45 @@ func marshalObjectIdentifier(oid []int) (ret []byte, err error) {
 }
 
-func marshalOID(oid string) ([]byte, error) {
+func oidSplitToIntArray(oid string, sep string) ([]int, error) {
```
> ```
> // oidSplitToIntArray is a helper function for marshalOID and takes an OID string
> // converts it to an array of integers and returns an int slice or an error
> ```
 comment added 
> Did you consider `oidToIntArray` ? Naming would be more consistent with the existing `oidToString` helper.
Yes, good idea. Renamed

TimRots replied:
Perfect, thank you.

TimRots commented:
> My tests are simple. The functions are called in a loop with one million iterations.
> In the case of oidToString, I additionally tested another implementation using a byte buffer, it is close to my result, but a little slower.
> For the oidToIntArray function, I took the marshalOID part, in which the string is directly converted into an array of numbers.
> You need to run it locally on your computer, The Go Playground does not allow you to run the program with a long timeout.
> 
> [Benchmark oidToIntArray](https://play.golang.org/p/cvujPO8-BEj)
> [Benchmark oidToString](https://play.golang.org/p/MQvzWmRVHi6)

Thank you for updating the PR and providing these tests @it1804. The performance improvements are impressive :)

```
marshalOID_Partial_Orig time: 3.713267932s
marshalOID_Partial_New time: 1.704001884s

oidToStringOrig time: 2.788383612s
oidToStringWithStringBuilder time: 182.199084ms
oidToStringWithBytesBuffer time: 227.554092ms
```

@SuperQ Could you give this PR a review please?

it1804 commented:
I am working on a couple more improvements, but I'm not sure of an equally significant improvement yet. I think if everything works out, I will make a new PR

it1804 commented:
@TimRots 
Could you run another test. I added calculating and setting the string buffer size to the oidToStringWithStringBuilderGrowed () function. It also greatly improves the performance in my production environment under high load. To calculate the length of the string oid from the array, an intermediate function was added parseOidLength

[Benchmark oidToString with string Growed](https://play.golang.org/p/O8RgVYz3WCp)

brian-brazil commented on helper.go:
```diff
@@ -463,41 +463,68 @@ func marshalObjectIdentifier(oid []int) (ret []byte, err error) {
+}
 
-	mOid, err := marshalObjectIdentifier(oidBytes)
+func marshalOID(oid string) ([]byte, error) {
```
I don't see how this is doing any marshalling, all this function effectively does is strip a leading period from what is already an oid.

it1804 replied:
I have not changed the names of the functions. I just changed part of the code from this function and moved it into a new function oidToIntArray () for readability.

brian-brazil replied:
That doesn't really address my point, if you care about efficiency you can make this function a lot simpler by deleting most of its code. The name can also be improved.

SuperQ replied:
@brian-brazil Can you be more explicit about what you mean by "deleting most of the code"?

brian-brazil replied:
Per above "all this function effectively does is strip a leading period from what is already an oid."

brian-brazil reviewed (commented):
Can you add micro-benchmarks, so we can see the change in allocs? I'd have expected all of this to be on the stack.

it1804 commented:
> Can you add micro-benchmarks, so we can see the change in allocs? I'd have expected all of this to be on the stack.

[bench_oid_to_int_array_test.go](https://play.golang.org/p/qbhCZ1PqU5D)
go test -bench=. -benchmem -benchtime=100000x bench_oid_to_int_array_test.go
Benchmark_marshalOID_Orig-24              100000              2296 ns/op             408 B/op          5 allocs/op
Benchmark_marshalOID_New-24               100000              1324 ns/op             208 B/op          3 allocs/op
PASS
ok      command-line-arguments  0.385s

[bench_oid_to_string_test.go](https://play.golang.org/p/8AUJofpysxj)
go test -bench=. -benchmem -benchtime=100000x bench_oid_to_string_test.go
Benchmark_oidToString_Orig-24                             100000              2444 ns/op             336 B/op         11 allocs/op
Benchmark_oidToString_WithStringBuilderGrowed-24          100000              1739 ns/op             144 B/op         10 allocs/op
Benchmark_oidToString_WithStringBuilder-24                100000              1962 ns/op             320 B/op         14 allocs/op
Benchmark_oidToString_WithBytesBuffer-24                  100000              2525 ns/op             352 B/op         12 allocs/op
PASS
ok      command-line-arguments  0.891s

efronlicht commented:
This is a big issue for my employer. We were able to come up with a solution that involves only one allocation.

I'm not sure if I can legally just give you our code, but what I _can_ do is give you a hint:

the algorithm you want looks something like this:

- count the size in bytes of our output. this should be the base-10 length of each section, +1 for each section.
(i.e, `[]int{23, 3}` -> `".23.3"` -> `"." + "23" + "." + "3"`
- allocate a buffer of `[]byte` exactly once.
- write to the buffer. `strconv` in the standard library should give you some hints on how to do this efficiently.
- convert the buffer to a string. this will do an additional allocation unless you use `unsafe`.

if you do it right, you should have only a single allocation, and it should run significantly faster. good luck!

TimRots commented:
> the algorithm you want looks something like this:
> 
>     * count the size in bytes of our output. this should be the base-10 length of each section, +1 for each section. this will involve one iteration through the list.
>       (i.e, `[]int{23, 3}` -> `".23.3"` -> `"." + "23" + "." + "3"`
> 
>     * allocate a buffer of `[]byte` with that size exactly once.
> 
>     * write the periods and base-10 formatted integers into the buffer. `strconv` in the standard library should give you some hints on how to do this efficiently. you should only need to iterate through one more time.
> 
>     * convert the buffer to a string. this will cost an additional allocation unless you use `unsafe`, but you can guarantee the buffer won't escape, so `unsafe` should be fine.
> 
> 
> if you do it right, you should have only a single allocation, and it should run significantly faster. good luck!

As much as I'm liking the described algorithm (& initiative of joining the conversation) @efronlicht. 'Righting' this at the cost of introducing ``unsafe`` in gosnmp is imho not preferred.
It's just too easy to introduce bugs which do not outweigh the benefit of this performance optimization.

If we are able to find a solution without ``unsafe`` that would be much welcomed however!

it1804 commented:
Without using unsafe in the oidToString () function, if this unsafe is removed, when using the byte array [] there will be 2 allocations in the test, but judging by my tests, the speed will not drop much. Fortunately, the size of the pre-allocated buffer is easy to calculate here. If you do not calculate the buffer size in advance, performance degradation will occur. In terms of speed, compared to my original StringBuilder and BytesBuffer, this is a big success. Thanks @efronlicht for the tip.

In another function that converts the oid string to asn1 byte string - marshalOID (), I also came up with an algorithm that avoids intermediate storage in [] int64, which will eliminate the oidToIntArray () function I wanted to add, and possibly marshalOID () itself instead of which I want to make marshalObjectIdentifier () the main and only one in converting oid string to asn1 byte string. I have not done all the tests yet, but there will be 2 allocations per operation beforehand if everything works out. In the worst case, three. I have not found an easy way to calculate the buffer size in advance without a double pass, because in base128 encoding, the size in bytes can range from 1 to 5 bytes.

In the next few days, I will update the commits with the new changes.

it1804 commented:
To my regret, the marshilling / unmarshalling that I wanted to fix does not allow me to achieve the performance that I achieved by commenting out all the lines related to logging. The difference reaches at least 20-30% in speed
Issue https://github.com/gosnmp/gosnmp/issues/313 is only part of a global problem that has not yet been resolved.

efronlicht commented:
> > the algorithm you want looks something like this:
> > ```
> > * count the size in bytes of our output. this should be the base-10 length of each section, +1 for each section. this will involve one iteration through the list.
> >   (i.e, `[]int{23, 3}` -> `".23.3"` -> `"." + "23" + "." + "3"`
> > 
> > * allocate a buffer of `[]byte` with that size exactly once.
> > 
> > * write the periods and base-10 formatted integers into the buffer. `strconv` in the standard library should give you some hints on how to do this efficiently. you should only need to iterate through one more time.
> > 
> > * convert the buffer to a string. this will cost an additional allocation unless you use `unsafe`, but you can guarantee the buffer won't escape, so `unsafe` should be fine.
> > ```
> > 
> > 
> > if you do it right, you should have only a single allocation, and it should run significantly faster. good luck!
> 
> As much as I'm liking the described algorithm (& initiative of joining the conversation) @efronlicht. 'Righting' this at the cost of introducing `unsafe` in gosnmp is imho not preferred.
> It's just too easy to introduce bugs which do not outweigh the benefit of this performance optimization.
> 
> If we are able to find a solution without `unsafe` that would be much welcomed however!

`unsafe` is just to skip the last alloc. you can simply convert the `[]byte` to `string ` directly and keep the rest the same and you'll maintain much of the savings.

efronlicht commented:
> I have not found an easy way to calculate the buffer size in advance without a double pass, because in base128 encoding, the size in bytes can range from 1 to 5 bytes.

Two hints:
- That's less than a word of size on a 64-bit machine, so there's no need to be cheap: just allocate two bytes on the stack and be done with it. `make([]byte, 0, 8)` should work well.
- There's no reason to be afraid of a double pass. Running through something in the memory cache is blazing fast.

> For some reason I thought that StringBuilder and BytesBuffer are more optimized for this.

They are very good general-purpose solutions, but if we know the specific bounds of our task, we can do better! In fact, reading the source code for them is a great way to learn how to efficiently handle memory.

Anyways, I'm working on permission to directly submit patches. Everyone here is willing, but things can move slowly in the corporate world.

Once again, good luck! Maintaining a big open-source project like this is hard work. Thank you for your time and attention!

it1804 commented:
> 
> 
> This is a big issue for my employer. We were able to come up with a solution that involves only one allocation.
> 
> I'm not sure if I can legally just give you our code, but what I _can_ do is give you a hint:
> 
> the algorithm you want looks something like this:
> 
>     * count the size in bytes of our output. this should be the base-10 length of each section, +1 for each section. this will involve one iteration through the list.
>       (i.e, `[]int{23, 3}` -> `".23.3"` -> `"." + "23" + "." + "3"`
> 
>     * allocate a buffer of `[]byte` with that size exactly once.
> 
>     * write the periods and base-10 formatted integers into the buffer. `strconv` in the standard library should give you some hints on how to do this efficiently. you should only need to iterate through one more time.
> 
>     * convert the buffer to a string. this will cost an additional allocation unless you use `unsafe`, but you can guarantee the buffer won't escape, so `unsafe` should be fine.
> 
> 
> if you do it right, you should have only a single allocation, and it should run significantly faster. good luck!

Please see my new version, it does not require calculating the slice size, but is set by a constant. Despite the buffer of the maximum possible size for the oid size, there are no unnecessary copy allocations. I suspect that the byte buffer in the case of the cap constant is created already at the compilation stage, so everything happens without additional copying of data in the heap.

TimRots commented:
> > This is a big issue for my employer. We were able to come up with a solution that involves only one allocation.
> > I'm not sure if I can legally just give you our code, but what I _can_ do is give you a hint:
> > the algorithm you want looks something like this:
> > ```
> > * count the size in bytes of our output. this should be the base-10 length of each section, +1 for each section. this will involve one iteration through the list.
> >   (i.e, `[]int{23, 3}` -> `".23.3"` -> `"." + "23" + "." + "3"`
> > 
> > * allocate a buffer of `[]byte` with that size exactly once.
> > 
> > * write the periods and base-10 formatted integers into the buffer. `strconv` in the standard library should give you some hints on how to do this efficiently. you should only need to iterate through one more time.
> > 
> > * convert the buffer to a string. this will cost an additional allocation unless you use `unsafe`, but you can guarantee the buffer won't escape, so `unsafe` should be fine.
> > ```
> > 
> > 
> > if you do it right, you should have only a single allocation, and it should run significantly faster. good luck!
> 
> Please see my new version, it does not require calculating the slice size, but is set by a constant. Despite the buffer of the maximum possible size for the oid size, there are no unnecessary copy allocations. I suspect that the byte buffer in the case of the cap constant is created already at the compilation stage, so everything happens without additional copying of data in the heap.
> Also I got rid of intermediate copies in [] int array and related functions marshalOID () and oidToString ()

Testing the updated PR yields the following results:
```
BenchmarkParseObjectIdentifierAsByteArray-4   	  400000	       181.1 ns/op	       0 B/op	       0 allocs/op
BenchmarkParseObjectIdentifierAsString-4      	  400000	       206.6 ns/op	      48 B/op	       1 allocs/op
BenchmarkMarshalObjectIdentifier-4            	  400000	       122.6 ns/op	       0 B/op	       0 allocs/op
BenchmarkSendOneRequest-4                     	  400000	       38394 ns/op	    2520 B/op	      83 allocs/op
```
vs current 1.31.1
```
BenchmarkOidToString-4      	  400000	       176.3 ns/op	     112 B/op	       2 allocs/op
BenchmarkSendOneRequest-4   	  400000	       39293 ns/op	    3256 B/op	      90 allocs/op
```
while using ``go test -bench=. -benchmem -benchtime=400000x *.go``

@SuperQ @brian-brazil Could we consider having this, based on the added benchmark info?

it1804 commented:
> 
> Testing the updated PR yields the following results:
> 
> ```
> BenchmarkParseObjectIdentifierAsByteArray-4   	  400000	       181.1 ns/op	       0 B/op	       0 allocs/op
> BenchmarkParseObjectIdentifierAsString-4      	  400000	       206.6 ns/op	      48 B/op	       1 allocs/op
> BenchmarkMarshalObjectIdentifier-4            	  400000	       122.6 ns/op	       0 B/op	       0 allocs/op
> BenchmarkSendOneRequest-4                     	  400000	       38394 ns/op	    2520 B/op	      83 allocs/op
> ```
> 
> vs current 1.31.1
> 
> ```
> BenchmarkOidToString-4      	  400000	       176.3 ns/op	     112 B/op	       2 allocs/op
> BenchmarkSendOneRequest-4   	  400000	       39293 ns/op	    3256 B/op	      90 allocs/op
> ```
> 
> while using `go test -bench=. -benchmem -benchtime=400000x *.go`
> 
> @SuperQ @brian-brazil Could we consider having this, based on the added benchmark info?

я выпустил новую версию патча, в котором полностью перешел на использование массива байт вместо bytes.Buffer.
Этот результат понравился мне больше.
Для сравнения go test -bench=. -benchmem -benchtime=400000x *.go:

Current master
`BenchmarkSendOneRequest-24        400000             80557 ns/op            3259 B/op         90 allocs/op
`

Previous patch:
`BenchmarkSendOneRequest-24                                400000             72618 ns/op            2522 B/op         83 allocs/op
`

New patch:
`BenchmarkSendOneRequest-24        400000             70879 ns/op            2114 B/op         79 allocs/op`

I also moved the buffer allocation for marshalLength () and parseObjectIdentifier () back inside these functions. In fact, in the full test, the increase in speed turned out to be small, plus such a code did not seem very nice.
But as it turned out, using a byte array everywhere turned out to be more efficient. 

PS:
And there is an added bonus, as I said, the registration code in the current implementation is the biggest performance issue. I have rewritten the code responsible for this, but have not fully tested it yet. Does it make sense to make these changes to this PR as well?

Test with disabled logging
`BenchmarkSendOneRequest-24                                400000             55143 ns/op            1360 B/op         39 allocs/op`

efronlicht commented:
P.S: consider just using build tags to enable and disable logging. The compiler is smart enough to optimize out functions which are literally no-ops.

it1804 commented:
> 
> 
> P.S: consider just using build tags to enable and disable logging. The compiler is smart enough to optimize out functions which are literally no-ops.

Compiler optimizations don't always actually work. Below is a test benchmark that proves this. I presented the Logger structure exactly as it is used now.
The first two benchmarks call logPrintf () and logPrint () in exactly the same way as in the original gosnmp code.
The last two marks before the call additionally check the loggingEnabled flag before calling these functions.
As you can see, checking the loggingEnabled flag inside logPrintf and logPrint is ineffective. But it is effective before calling from the outside.
You can verify this yourself by commenting out one or more logPrintf () and logPrint () in the library code.

```
Benchmark_logPrintf-24                            100000               223.7 ns/op            16 B/op          1 allocs/op
Benchmark_logPrint-24                             100000               101.2 ns/op            16 B/op          1 allocs/op
Benchmark_logPrintf_With_External_Flag-24         100000                 1.260 ns/op           0 B/op          0 allocs/op
Benchmark_logPrint_With_External_Flag-24          100000                 1.452 ns/op           0 B/op          0 allocs/op

```

```
//go test -bench=. -benchmem -benchtime=100000x logger_test.go
package main

import (
        "io/ioutil"
        "log"
        "testing"
)

type GoSNMP struct {
        Logger         Logger
        loggingEnabled bool
}

type Logger interface {
        Print(v ...interface{})
        Printf(format string, v ...interface{})
}

func (x *GoSNMP) logPrint(v ...interface{}) {
        if x.loggingEnabled {
                x.Logger.Print(v...)
        }
}

func (x *GoSNMP) logPrintf(format string, v ...interface{}) {
        if x.loggingEnabled {
                x.Logger.Printf(format, v...)
        }
}

func Benchmark_logPrintf(b *testing.B) {
        g := &GoSNMP{
                Logger: log.New(ioutil.Discard, "", 0),
        }
        g.loggingEnabled = false

        for i := 0; i < b.N; i++ {
                g.logPrintf("Test %v", g.loggingEnabled)
        }
}

func Benchmark_logPrint(b *testing.B) {
        g := &GoSNMP{
                Logger: log.New(ioutil.Discard, "", 0),
        }
        g.loggingEnabled = false

        for i := 0; i < b.N; i++ {
                g.logPrint("Test")
        }
}

func Benchmark_logPrintf_With_External_Flag(b *testing.B) {
        g := &GoSNMP{
                Logger: log.New(ioutil.Discard, "", 0),
        }
        g.loggingEnabled = false

        for i := 0; i < b.N; i++ {
                if g.loggingEnabled {
                        g.logPrintf("Test %v", g.loggingEnabled)
                }
        }
}

func Benchmark_logPrint_With_External_Flag(b *testing.B) {
        g := &GoSNMP{
                Logger: log.New(ioutil.Discard, "", 0),
        }
        g.loggingEnabled = false

        for i := 0; i < b.N; i++ {
                if g.loggingEnabled {
                        g.logPrint("Test")
                }
        }
}

```

efronlicht commented:
> ally work. Below is a test benchmark that proves this. I presented the Logger structure exactly as it is used now.

You don't understand what I'm talking about.

My suggestion is as follows: use build tags to have two separate versions of the methods `.logPrint` and `.logPrintf`. Without the build tag, have those functions be no-ops: as in, they literally have no function body. This lets you eliminate the `loggingEnabled` flag entirely, should you choose, since it's now a compile-time check.

if you want proof that the compiler will optimize it out, I suggest you compare the generated assembly:

[old-style](https://godbolt.org/z/Gj1j8MY7c)
[my suggestion](https://godbolt.org/z/zY46en3Pq)

SuperQ commented:
@efronlicht Would you mind opening a PR to implement the build tag options?

SuperQ commented:
@it1804 In the case of build tags, this optimization will work just fine. It's more like a C `ifdef`.

efronlicht commented:
> @efronlicht Would you mind opening a PR to implement the build tag options?

I'm not sure if I legally can. It might be considered breach of contract by my employer. I'm hoping I can get official permission in a few weeks: these things move slowly.

it1804 commented:
> 
> 
> > Compiler optimizations don't always actually work. Below is a test benchmark that proves this. I presented the Logger structure exactly as it is used now.
> 
> My suggestion is as follows: use build tags to have two separate versions of the methods `.logPrint` and `.logPrintf`. Without the build tag, have those functions be no-ops: as in, they literally have no function body. This lets you eliminate the `loggingEnabled` flag entirely, should you choose, since it's now a compile-time check.
> 
> if you want proof that the compiler will optimize it out, I suggest you compare the generated assembly: here are links to the godbolt compiler explorer for your convenience.
> 
> [old-style](https://godbolt.org/z/Gj1j8MY7c)
> [my suggestion](https://godbolt.org/z/zY46en3Pq)
> 
> While moving the runtime check outside will improve performance, it will add a ton of extra lines to the codebase. This isn't the worst thing in the world, but I think build tags are the simplest and most performant solution.
> 
> In any case, its your call. You guys run the project, not me - just trying to help out upstream where I can.

Using assembly tags is the best option for me, we usually do this in C ++.
But I don't understand how to maintain a library with conditional compilation directives if it will be distributed as a module. In version 1.16, the modular system is already enabled by default. And not everyone builds the required libraries locally, they just run "go get".
Yes, what I suggested would add some coding problems, but I thought it would be better for the community, downloaded by dependencies via "go get" and no additional manipulation.
I'm a bit new to the world of Go, and if you tell me how to handle this without rebuilding locally, with or without modules, I can adapt the code.
In my assembly, everything related to logging is cut out to speed up the launch of the project. I just wanted to draw your attention to this problem.

efronlicht commented:
> I understood everything, thank you very much for the idea! It was difficult to maintain two builds, with and without logging. But I'm a bit new to Go, and I might miss some points. Perhaps I will try to implement it.

Here's a minimal example of how build tags work:  [gitlab.com/efronlicht/build-tags-example](https://gitlab.com/efronlicht/build-tags-example). This should give you everything you need to implement it. Good luck!

it1804 commented:
But I still think the Logger class needs refactoring. 
This is how I wanted to implement: 
```
type LoggerInterface interface {
    Print(v ...interface{})
    Printf(format string, v ...interface{})
}

type Logger struct {
    Enabled bool
    logger LoggerInterface
}
```
In the case of using build tags, the Enabled variable becomes unnecessary, you can exclude it.

And install the logger through a separate function SetLogger () instead of directly assigning to the Logger variable.

Then the debug functions can be always called via x.Logger.Printf (...) or x.Logger.Print (...) only

it1804 commented:
I propose to move away from logging theme in this. If I can, then I will open a new pr, and on this I will try to finish what I started earlier, taking into account new information for me.
I am concerned about the modifications that I have already proposed in the patch in terms of code cleanliness. In fact, I tried to bind the intermediate variables into one long array during marshaling so that the memory allocation for each function can be minimized. Unfortunately this was not always successful and temporary arrays are still needed in some places.
I am also worried about the implementation of ParseObjectIdentifier () / MarshalObjectIdentifier ()

TimRots commented:
A lot of great ideas are shared in this PR however kindly asking to move the discussion on logger improvements into #313 or a separate issue. @it1804 If any help is needed on that hit me up on slack or per email.

For now, can we keep the PR as originally intended with (measurable) improvements to ``marshalOID()`` and ``oidToString()`` functions. If more improvements can be made we can iterate on that later.

it1804 commented:
> 
> 
> A lot of great ideas are shared in this PR however kindly asking to move the discussion on logger improvements into #313 or a separate issue. @it1804 If any help is needed I'm glad to assist.
> 
> For now, can we keep the PR as originally intended with (measurable) improvements to `marshalOID()` and `oidToString()` functions. If more improvements can be made we can iterate on that later.

Please see my solution to the logging problem https://github.com/gosnmp/gosnmp/pull/329

it1804 commented:
I want to clarify this PR.
In my tests, it turned out that using one large byte buffer is more optimal than allocating its own buffer inside each function, and then copying it to the main one. So whenever possible I tried to chain the marshaling functions together using one big byte array.
Initially, I also considered the option of pre-calculating the size of each buffer to avoid unnecessary memory allocation.
But, firstly, it is not always possible, and secondly, as tests have shown, it is not very effective. In many cases, pre-calculating the length of an array takes longer than simply creating a zero-length array.

it1804 commented:
I didn't like my solution with replacing the byte buffer with a simple array. The code turned out to be confusing and inconvenient to maintain.
I simplified the commit and decided to keep only the marshalObjectIdentifier () and parseObjectIdentifier () refactorings.
Also, on the 386 platform, a number overflow occurred if the oid contained the value of the Max oid sub-identifier 4294967295 (https://tools.ietf.org/html/rfc2578#section-7.1.3). Solved by replacing int with int64 in parseObjectIdentifier () and parseBase128Int ()

The result is slightly worse than in the first version 

New Version based on 1.31 with logging enabled
go test -v -bench=. -benchmem -benchtime=100000x -tags all
```
BenchmarkParseObjectIdentifier
BenchmarkParseObjectIdentifier-24         100000               755.1 ns/op            88 B/op          2 allocs/op
BenchmarkMarshalObjectIdentifier
BenchmarkMarshalObjectIdentifier-24       100000               627.6 ns/op           112 B/op          2 allocs/op
BenchmarkSendOneRequest
BenchmarkSendOneRequest-24                100000             67702 ns/op            2528 B/op         80 allocs/op
```

New Version based on 1.31 with logging disabled
go test -v -bench=. -benchmem -benchtime=100000x -tags all,gosnmp_nodebug
```
BenchmarkParseObjectIdentifier
BenchmarkParseObjectIdentifier-24         100000               677.8 ns/op            88 B/op          2 allocs/op
BenchmarkMarshalObjectIdentifier
BenchmarkMarshalObjectIdentifier-24       100000               691.8 ns/op           112 B/op          2 allocs/op
BenchmarkSendOneRequest
BenchmarkSendOneRequest-24                100000             62031 ns/op            2032 B/op         49 allocs/op

```

TimRots reviewed (approved):
Thanks @it1804. I can approve these changes as-is.

@SuperQ Could you please give this PR a final review?

SuperQ reviewed (approved):
LGTM

SuperQ commented:
@brian-brazil Would you mind helping with a final review?

brian-brazil commented on helper.go:
```diff
@@ -442,62 +439,59 @@ func marshalLength(length int) ([]byte, error) {
+		var val int64 = 0
+		for j < oidLength && oid[j] != '.' {
+			ch := int64(oid[j] - '0')
+			if ch > 9 {
```
Or `< 0` ?

it1804 replied:
It's not necessary. The original byte at the index position (oid[j]) is of type uint8. The expression oid [j] - '0' will also always be > 0.

brian-brazil reviewed (commented):
Which publicly-visible interfaces does this change?

it1804 commented:
> 
> 
> Which publicly-visible interfaces does this change?

The fix removes the marshalOID() and oidToString() internal helper functions.
Instead, marshalObjectIdentifier() and parseObjectIdentifier() are now used directly without intermediate conversions.

Public interfaces are not affected.

brian-brazil commented:
:+1:

TimRots commented:
Thank you for the reviews.

@it1804 could you please add an entry to the ``CHANGELOG.md`` mentioning
```
* [ENHANCEMENT]: helper.go: Improved OID marshaling with sub-identifier validation as per rfc2578 section-3.5 #321
```
then we are ready to have this merged :)
